### PR TITLE
Use dask as new parallel backend

### PR DESCRIPTION
**Pull request by mfeurer**
This PR makes use of the latest SMAC developments and uses Dask as a backend to execute runs in parallel. More specifically:

* dask is used as a parallel backend
* n_jobs now spawns a dask cluster automatically
* alternatively, a dask client can be passed by the user
* examples are updated to reflect this change
* drop the `shared_mode`

### CODE REVIEW

File: autosklearn/__init__.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 requirements = pkg_resources.resource_string('autosklearn', 'requirements.txt')
 requirements = requirements.decode('utf-8')
 
-dependencies.verify_packages(requirements)
+# dependencies.verify_packages(requirements)
```
Comment (franchuterivera) [inline autosklearn/__init__.py:13]: Do you want to leave this line commented out?

File: autosklearn/automl.py [hunk 2]
```diff
@@ -110,7 +111,8 @@ def __init__(self,
                  exclude_preprocessors=None,
                  resampling_strategy='holdout-iterative-fit',
                  resampling_strategy_arguments=None,
-                 shared_mode=False,
+                 n_jobs=None,
```
Comment (franchuterivera) [inline autosklearn/automl.py:114]: As shared mode is going away, we can also remove this line (from below, I found it on a grep on a repo)?
https://github.com/automl/auto-sklearn/blob/00ecf0783512e95c46459f7e4a084ea805cffb39/test/test_automl/test_automl.py#L44

File: autosklearn/automl.py [hunk 3]
```diff
@@ -386,16 +379,16 @@ def fit(
...
-            if not requirement:
-                continue
-            match = RE_PATTERN.match(requirement)
-            if match:
-                name = match.group('name')
-                module_dist = pkg_resources.get_distribution(name)
-                self._logger.debug('  %s', module_dist)
-            else:
-                raise ValueError('Unable to read requirement: %s' % requirement)
+        # for requirement in requirements:
```
Comment (franchuterivera) [inline autosklearn/automl.py:382]: Same here with the comments :), yeah I know this is a draft, so probably something you already had in mind removing.

File: autosklearn/automl.py [hunk 4]
```diff
@@ -421,7 +414,8 @@ def fit(
         self._logger.debug('  resampling_strategy: %s', str(self._resampling_strategy))
         self._logger.debug('  resampling_strategy_arguments: %s',
                            str(self._resampling_strategy_arguments))
-        self._logger.debug('  shared_mode: %s', str(self._shared_mode))
+        self._logger.debug('  n_jobs: %s', str(self._n_jobs))
+        self._logger.debug('  dask_client: %s', str(self._dask_client))
```
Comment (franchuterivera) [inline autosklearn/automl.py:418]: Do You think it also make sense to print the scheduler info? Here is the comparison:
```
>>> c = dask.distributed.Client()
>>> print(c)
<Client: 'tcp://127.0.0.1:40415' processes=4 threads=12, memory=16.69 GB>
```
vs
```
>>> print(c.scheduler_info())
{'type': 'Scheduler', 'id': 'Scheduler-6120002b-f080-4c31-a217-67cfea0be3b8', 'address': 'tcp://127.0.0.1:40415', 'services': {'dashboard': 8787}, 'workers': {'tcp://127.0.0.1:35037': {'type': 'Worker', 'id': 2, 'host': '127.0.0.1', 'resources': {}, 'local_directory': '/home/chico/work/auto-sklearn_prdask/dask-worker-space/dask-worker-space/worker-ret8pbbf', 'name': 2, 'nthreads': 3, 'memory_limit': 4171328512, 'last_seen': 1600200774.5238085, 'services': {'dashboard': 43551}, 'metrics': {'cpu': 6.0, 'memory': 52838400, 'time': 1600200774.516168, 'read_bytes': 4357.279003306836, 'write_bytes': 3565.2278291709317, 'num_fds': 24, 'executing': 0, 'in_memory': 0, 'ready': 0, 'in_flight': 0, 'bandwidth': {'total': 100000000, 'workers': {}, 'types': {}}}, 'nanny': 'tcp://127.0.0.1:45709'}, 'tcp://127.0.0.1:40489': {'type': 'Worker', 'id': 3, 'host': '127.0.0.1', 'resources': {}, 'local_directory': '/home/chico/work/auto-sklearn_prdask/dask-worker-space/dask-worker-space/worker-tn8ppdc0', 'name': 3, 'nthreads': 3, 'memory_limit': 4171328512, 'last_seen': 1600200774.523963, 'services': {'dashboard': 34521}, 'metrics': {'cpu': 6.0, 'memory': 52699136, 'time': 1600200774.5164387, 'read_bytes': 4009.613552863556, 'write_bytes': 3214.8973369440196, 'num_fds': 24, 'executing': 0, 'in_memory': 0, 'ready': 0, 'in_flight': 0, 'bandwidth': {'total': 100000000, 'workers': {}, 'types': {}}}, 'nanny': 'tcp://127.0.0.1:32925'}, 'tcp://127.0.0.1:41667': {'type': 'Worker', 'id': 0, 'host': '127.0.0.1', 'resources': {}, 'local_directory': '/home/chico/work/auto-sklearn_prdask/dask-worker-space/dask-worker-space/worker-i6y9l53r', 'name': 0, 'nthreads': 3, 'memory_limit': 4171328512, 'last_seen': 1600200774.5185788, 'services': {'dashboard': 35937}, 'metrics': {'cpu': 4.0, 'memory': 52658176, 'time': 1600200774.515757, 'read_bytes': 4008.394923610346, 'write_bytes': 3213.9202432941665, 'num_fds': 24, 'executing': 0, 'in_memory': 0, 'ready': 0, 'in_flight': 0, 'bandwidth': {'total': 100000000, 'workers': {}, 'types': {}}}, 'nanny': 'tcp://127.0.0.1:44841'}, 'tcp://127.0.0.1:42781': {'type': 'Worker', 'id': 1, 'host': '127.0.0.1', 'resources': {}, 'local_directory': '/home/chico/work/auto-sklearn_prdask/dask-worker-space/dask-worker-space/worker-uzr_gyvv', 'name': 1, 'nthreads': 3, 'memory_limit': 4171328512, 'last_seen': 1600200774.5124593, 'services': {'dashboard': 43121}, 'metrics': {'cpu': 4.0, 'memory': 52826112, 'time': 1600200774.0123239, 'read_bytes': 12385.847500842423, 'write_bytes': 12117.59090392746, 'num_fds': 24, 'executing': 0, 'in_memory': 0, 'ready': 0, 'in_flight': 0, 'bandwidth': {'total': 100000000, 'workers': {}, 'types': {}}}, 'nanny': 'tcp://127.0.0.1:46601'}}}

```

File: autosklearn/__init__.py (hunk 1 above)
Comment (mfeurer) [reply autosklearn/__init__.py:13]: No, the issue is that the dependency parser cannot cope with the `git+https://github.com/automl/smac3@development`. Therefore it's temporarily commented out.

File: autosklearn/automl.py (hunk 3 above)
Comment (mfeurer) [reply autosklearn/automl.py:382]: Yes, same issue as above. Maybe I should just update the dependency parser to ignore the version in such cases.

File: autosklearn/automl.py (hunk 4 above)
Comment (mfeurer) [reply autosklearn/automl.py:418]: I'm not sure. While running, the scheduler information will be available as a scheduler file in machine readable format anyway (https://github.com/automl/SMAC3/blob/development/smac/tae/dask_runner.py#L115), so I'm not 100% sure if there's additional helpful information here. If there's any, please let me know.

**Comment by franchuterivera** [comment]
Hello Matthias, BTW, I was playing around with this code to test the dask ensemble builder, but just in case you have not seen this issue:
```
[WARNING] [2020-09-18 19:12:52,253:smac.optimizer.smbo.SMBO] Exhausted configuration budget
[WARNING] [2020-09-18 19:12:52,253:smac.optimizer.smbo.SMBO] Exhausted configuration budget
auto-sklearn results:
  Dataset name: breast_cancer
  Metric: accuracy
  Number of target algorithm runs: 83
  Number of successful target algorithm runs: 0
  Number of crashed target algorithm runs: 83
  Number of target algorithms that exceeded the time limit: 0
  Number of target algorithms that exceeded the memory limit: 0
```

All configurations are crashing. BTW, the backend object is deleted in my ensemble builder, when a run is complete. Do you think it makes sense to move the __del__ of the backend to the estimator (i think in the smbo you faced the same problem)?

**Comment by mfeurer** [comment]
> Hello Matthias, BTW, I was playing around with this code to test the dask ensemble builder, but just in case you have not seen this issue:

I cannot reproduce this on my machine. I just tried the `example_parallel_n_jobs` and it works well. Did you guard the code with `__name__ == '__main__'`?

> BTW, the backend object is deleted in my ensemble builder, when a run is complete. Do you think it makes sense to move the del of the backend to the estimator (i think in the smbo you faced the same problem)?

Yes, I think that would make sense.

**Pull request merged**